src/axcrypt_fmt_plug.c

```diff
@@ -0,0 +1,354 @@
+            + atoi16[ARCH_INDEX(p[i * 2 + 1])];
+
+    /* if key-file present */
+    if ((p = strtokm(NULL, "*")) != NULL){
+        cs.keyfile = (char*) mem_calloc(strlen(p)/2, sizeof(char));
+        for (i = 0; i < strlen(p)/2; i++)
+            cs.keyfile[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16
+                + atoi16[ARCH_INDEX(p[i * 2 + 1])];
+    }
+    MEM_FREE(keeptr);
+
+    cs.dsalt.salt_cmp_offset = SALT_CMP_OFF(struct custom_salt, salt);
+    cs.dsalt.salt_cmp_size = SALT_CMP_SIZE(struct custom_salt, salt, wrappedkey, 0);
+    cs.dsalt.salt_alloc_needs_free = 0;
+
+    ptr = mem_alloc_tiny(sizeof(struct custom_salt), MEM_ALIGN_WORD);
+    memcpy(ptr, &cs, sizeof(struct custom_salt));
+
+    return (void *) &ptr;
+}
+
+static void set_salt(void *salt)
+{
+    cur_salt = *(struct custom_salt **) salt;
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+    const int count = *pcount;
+    int index = 0;
+
+    if (any_cracked) {
+        memset(cracked, 0, cracked_size);
+        any_cracked = 0;
+    }
+
+#ifdef _OPENMP
+#pragma omp parallel for
+    for (index = 0; index < count; index++)
+#endif
+    {
+        /*
+            NUMBER_AES_BLOCKS = 2
+            AES_BLOCK_SIZE = 16
+        */
+
+        unsigned char KEK[20], lsb[24], cipher[16];
+        AES_KEY akey;
+        SHA_CTX ctx;
+
+        int i, j, nb_iterations = cur_salt->key_wrapping_rounds;
+
+        SHA1_Init(&ctx);
+        SHA1_Update(&ctx, (unsigned char *) saved_key[index], 
+                    strlen(saved_key[index]));
+        /* if key-file provided */
+        if (cur_salt->keyfile != NULL)
+            SHA1_Update(&ctx, (unsigned char *) cur_salt->keyfile, 
+                        strlen(cur_salt->keyfile));
+        SHA1_Final( KEK, &ctx );
+
+        /* hash XOR salt => KEK */
+        for (i = 0; i < sizeof(cur_salt->salt); i++)
+            KEK[i] ^= cur_salt->salt[i];
+
+        memcpy(lsb, cur_salt->wrappedkey + 8, 16);
+
+        memset(&akey, 0, sizeof(AES_KEY));
+        AES_set_decrypt_key(KEK, 128, &akey);
+
+        /* set msb */
+        memcpy(cipher, cur_salt->wrappedkey, 8);
+
+        /* custom AES un-wrapping loop */
+        for (j = nb_iterations - 1; j >= 0; j--) {
+
+            /* 1st block treatment */
+            /* MSB XOR (NUMBER_AES_BLOCKS * j + i) */
+            PUT_64BITS_XOR_MSB(cipher, 2 * j + 2);
+            /* R[i] */
+            memcpy(cipher + 8, lsb + 8, 8);
+            /* AES_ECB(KEK, (MSB XOR (NUMBER_AES_BLOCKS * j + i)) | R[i]) */
+            AES_decrypt(cipher, cipher, &akey);
+            memcpy(lsb + 8, cipher + 8, 8);
+
+            /* 2nd block treatment */
+            PUT_64BITS_XOR_MSB(cipher, 2 * j + 1);
+            memcpy(cipher + 8, lsb, 8);
+            AES_decrypt(cipher, cipher, &akey);
+            memcpy(lsb, cipher + 8, 8);
+        }
+        if (!memcmp(cipher, AES_WRAPPING_IV, 8)) {
+            if (cur_salt->keyfile != NULL)
+                MEM_FREE(cur_salt->keyfile);
```

Comment:
This use of free() on a global pointer within a parallel block is probably dangerous, but isn't the cause of the failures we're seeing.